### PR TITLE
[SPARK-32816][SQL] Fix analyzer bug when aggregating multiple distinct DECIMAL columns

### PR DESCRIPTION
**Pull request by linhongliu-db**
### What changes were proposed in this pull request?
This PR fixes a conflict between `RewriteDistinctAggregates` and `DecimalAggregates`.
In some cases, `DecimalAggregates` will wrap the decimal column to `UnscaledValue` using
different rules for different aggregates.

This means, same distinct column with different aggregates will change to different distinct columns
after `DecimalAggregates`. For example:
`avg(distinct decimal_col), sum(distinct decimal_col)` may change to
`avg(distinct UnscaledValue(decimal_col)), sum(distinct decimal_col)`

We assume after `RewriteDistinctAggregates`, there will be at most one distinct column in aggregates,
but `DecimalAggregates` breaks this assumption. To fix this, we have to switch the order of these two
rules.

### Why are the changes needed?
bug fix

### Does this PR introduce _any_ user-facing change?
no

### How was this patch tested?
added test cases

### CODE REVIEW

File: sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala [hunk 1]
```diff
@@ -2555,6 +2555,19 @@ class DataFrameSuite extends QueryTest
...
     checkAnswer(df.select($"pos" > $"neg"), Row(false))
   }
+
+  test("SPARK-32816: aggregating multiple distinct DECIMAL columns") {
+    withTempPath { path =>
+      spark.range(0, 100, 1, 1)
+        .selectExpr("id", "cast(id as decimal(9, 0)) as decimal_col")
+        .write.mode("overwrite")
+        .parquet(path.getAbsolutePath)
+      spark.read.parquet(path.getAbsolutePath).createOrReplaceTempView("test_table")
```
Comment (TJX2014) [inline sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala:2565]: Seems we need not to write parquet.

Comment (TJX2014) [reply sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala:2565]: `val df = spark.range(0, 50000, 1, 1).selectExpr("id", "cast(id as decimal(9, 0)) as ss_ext_list_price")
df.createOrReplaceTempView("test_table")
sql("select avg(distinct ss_ext_list_price), sum(distinct ss_ext_list_price) from test_table").explain` seems enough to reproduce.

**Comment by linhongliu-db** [comment]
@cloud-fan

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 2]
```diff
@@ -196,6 +195,8 @@ abstract class Optimizer(catalogManager: CatalogManager)
       EliminateSorts) :+
     Batch("Decimal Optimizations", fixedPoint,
       DecimalAggregates) :+
+    Batch("Distinct Aggregate Rewrite", Once,
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:198]: can we add a comment to say: this batch must be run after "Decimal Optimizations", as "Decimal Optimizations" may change the aggregate distinct column?

Comment (gatorsmile) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:198]: In the comment, we can also point out RewriteDistinctAggregates should be executed after the rule DecimalAggregates

File: sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala [hunk 3]
```diff
@@ -2555,6 +2555,15 @@ class DataFrameSuite extends QueryTest
     val df = Seq(0.0 -> -0.0).toDF("pos", "neg")
     checkAnswer(df.select($"pos" > $"neg"), Row(false))
   }
+
+  test("SPARK-32816: aggregating multiple distinct DECIMAL columns") {
+    spark.range(0, 100, 1, 1)
+      .selectExpr("id", "cast(id as decimal(9, 0)) as decimal_col")
+      .createOrReplaceTempView("test_table")
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala:2562]: nit: wrap the test with `withTempView`

Comment (maropu) [reply sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala:2562]: How about writing it like this w/o a temp view;
```
  test("SPARK-32816: aggregating multiple distinct DECIMAL columns") {
    checkAnswer(
      sql(
        s"""
           |SELECT AVG(DISTINCT decimal_col), SUM(DISTINCT decimal_col)
           |  FROM VALUES (CAST(1 AS DECIMAL(9, 0))) t(decimal_col)
         """.stripMargin),
      Row(XXX, XXX))
  }
```

Comment (maropu) [reply sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala:2562]: nit: Also, could you move this test into `SQLQueryTestSuite`?

File: sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala [hunk 4]
```diff
@@ -2555,6 +2555,17 @@ class DataFrameSuite extends QueryTest
     val df = Seq(0.0 -> -0.0).toDF("pos", "neg")
     checkAnswer(df.select($"pos" > $"neg"), Row(false))
   }
+
+  test("SPARK-32816: aggregating multiple distinct DECIMAL columns") {
+    withTempView("test_table") {
+      spark.range(0, 100, 1, 1)
+        .selectExpr("id", "cast(id as decimal(9, 0)) as decimal_col")
+        .createOrReplaceTempView("test_table")
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala:2563]: can you follow https://github.com/apache/spark/pull/29673#discussion_r485977416 ? That's a better idea.

File: sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala (hunk 3 above)
Comment (linhongliu-db) [reply sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala]: let me try this, I'm not sure if literal will behave differently

File: sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala (hunk 4 above)
Comment (linhongliu-db) [reply sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala:2563]: sure

File: sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala [hunk 5]
```diff
@@ -2555,6 +2555,16 @@ class DataFrameSuite extends QueryTest
...
   }
+
+  test("SPARK-32816: aggregating multiple distinct DECIMAL columns") {
+    checkAnswer(
+      sql(
+        s"""
+           |SELECT AVG(DISTINCT decimal_col), SUM(DISTINCT decimal_col)
+           |  FROM VALUES (CAST(1 AS DECIMAL(9, 0))) t(decimal_col)
+        """.stripMargin),
+      Row(1, 1))
```
Comment (maropu) [inline sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala:2566]: nit: the test `group-by.sql` looks enough for this issue, so could you remove this?

**Review by maropu — APPROVED** [review]
LGTM excep for one minor comment.

File: sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala (hunk 5 above)
Comment (linhongliu-db) [reply sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala:2566]: sorry, I forgot to remove this.

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Pull request closed without merging**